### PR TITLE
Use a pull model to withdraw deposit ETH balances

### PR DESCRIPTION
**Pull request by NicholasDotSol**
Push models have security and stability concerns, a pull model allows for better security, as well as more flexibility and freedom to third-party contract fallbacks given `.call.value()`'s gas-forwarding.

- Accumulated Eth balance can only be withdrawn after the contract reaches an end-state

### CODE REVIEW

File: solidity/contracts/deposit/DepositUtils.sol [hunk 1]
```diff
@@ -396,6 +399,26 @@ library DepositUtils {
...
+    /// @notice     Adds a given amount to an addresses withdrawal allowance
+    /// @dev        Withdrawals can only happen post contract-end-state.
+    function enableWithdrawal(DepositUtils.Deposit storage _d, address _withdrawer, uint256 _amount) internal {
+        withdrawalAllowances[_withdrawer].add(_amount);
+    }
+
+    /// @notice     Withdraw caller's allowance.
+    /// @dev        Withdrawals can only happen post contract-end-state.
+    function withdrawFunds(DepositUtils.Deposit storage _d) internal {
+        require(_d.inEndState(), "Contact not yet terminated");
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol:412]: Contract*

File: solidity/contracts/deposit/DepositLiquidation.sol [hunk 2]
```diff
@@ -190,21 +196,18 @@ library DepositLiquidation {
...
         // we simply ignore such a tiny amount and leave some wei dust in escrow
         uint256 contractEthBalance = address(this).balance;
-        address payable initiator = _d.liquidationInitiator;
 
-        if (initiator == address(0)){
-            initiator = address(0xdead);
-        }
-        if (contractEthBalance > 1) {
+        if (contractEthBalance > _valueToDistribute) {
+            uint256 remainingUnallocated = contractEthBalance.sub(_valueToDistribute);
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositLiquidation.sol:201]: `_valueToDistribute` really shouldn't start with an underscore, as a side note.

File: solidity/.soliumrc.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
     ],
     "rules": {
         "security/no-block-members": "off",
-        "security/no-inline-assembly": "off"
+        "security/no-inline-assembly": "off",
+        "security/no-call-value": "off"
```
Comment (Shadowfiend) [inline solidity/.soliumrc.json:9]: We don't need this.

File: solidity/contracts/deposit/DepositFunding.sol [hunk 4]
```diff
@@ -97,14 +90,13 @@ library DepositFunding {
...
-        /* solium-disable-next-line security/no-send */
-        _d.depositOwner().send(_d.keepSetupFee);
-        _d.pushFundsToKeepGroup(_seized.sub(_d.keepSetupFee));
-
         _d.setFailedSetup();
         _d.logSetupFailed();
-
         fundingTeardown(_d);
+
+        /* solium-disable-next-line security/no-send */
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositFunding.sol:97]: We can lose all of these.

File: solidity/contracts/deposit/DepositUtils.sol [hunk 5]
```diff
@@ -396,6 +399,26 @@ library DepositUtils {
...
+        require(withdrawalAllowances[msg.sender] > 0, "Nothing to withdraw"");
+
+        // zero-out to prevent reentrancy
+        uint256 available = withdrawalAllowances[msg.sender];
+        withdrawalAllowances[msg.sender] = 0;
+    
+        msg.sender.call.value(available)("");
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol:419]: We *will* need a linter annotation here I guess.

File: solidity/.soliumrc.json (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/.soliumrc.json:9]: you'd rather have the solium-disable comment?

Comment (Shadowfiend) [reply solidity/.soliumrc.json:9]: What I really want is a way to annotate a function as “I know this is going to do a `.call.value` and I'm completely okay with that”, and only in the call path of that one function is anyone allowed to do `.call.value`. I would like this compiler-enforced.

In the absence of such a guard, this at least draws extra attention to where we're doing `call.value`. But I wish we could annotate the whole function.

Basically I'm trying to use a linter to do a compiler's job and it's terrible. But I think it's marginally less terrible than not doing so.

Comment (NicholasDotSol) [reply solidity/.soliumrc.json:9]: just saw https://github.com/keep-network/tbtc/pull/569#discussion_r405896462

Comment (NicholasDotSol) [reply solidity/.soliumrc.json:9]: Makes sense to use linter annotation over the rule if it's just one instance

**Comment by NicholasDotSol** [comment]
close in favor of #570
